### PR TITLE
l10n-bumper - add capability to poll hg pushlog for latest revision info

### PR DESCRIPTION
**Pull request by escapewindow**
Elmo is slated to be retired. Instead of getting signoffs in Elmo and
taking the latest signed off revisions as the revisions in
l10n-changesets, let's poll hg pushlog for the latest revision unless
the `pin` boolean is set in the locale.

### CODE REVIEW

**Comment by escapewindow** [comment]
I'm still running tests, but I think this is ready for a review pass.

File: treescript/src/treescript/l10n.py [hunk 1]
```diff
@@ -77,26 +81,82 @@ def build_platform_dict(bump_config, repo_path):
...
+
+    Args:
+        locale (str): the locale to query
+        url (str): the [templatized] pushlog url
+
+    Returns:
+        tuple (locale, revision)
+
+    """
+    url = url % {"locale": locale}
```
Comment (bhearsum) [inline treescript/src/treescript/l10n.py:96]: nit: this should probably use `.format`

File: treescript/src/treescript/l10n.py [hunk 2]
```diff
@@ -77,26 +81,82 @@ def build_platform_dict(bump_config, repo_path):
...
+    with tempfile.NamedTemporaryFile() as fp:
```
Comment (bhearsum) [inline treescript/src/treescript/l10n.py:97]: nit: is there a reason this has to go to disk instead of just keeping it in memory? size?

File: treescript/src/treescript/l10n.py [hunk 3]
```diff
@@ -77,26 +81,82 @@ def build_platform_dict(bump_config, repo_path):
...
+        path = fp.name
+        await retry_async(download_file, args=(url, path), retry_exceptions=(DownloadError,))
+        revision_info = load_json_or_yaml(path, is_path=True)
+    last_push_id = revision_info["lastpushid"]
+    revision = revision_info["pushes"][str(last_push_id)]["changesets"][0]
```
Comment (bhearsum) [inline treescript/src/treescript/l10n.py:102]: Are these pushes guaranteed to only have one changeset? It might be safer to take `[-1]` instead of `[0]`.

File: treescript/src/treescript/l10n.py [hunk 4]
```diff
@@ -77,26 +81,82 @@ def build_platform_dict(bump_config, repo_path):
...
+                value["pin"] = old_contents[locale]["pin"]
+            else:
+                tasks.append(asyncio.create_task(semaphore_wrapper(semaphore, get_latest_revision(locale, bump_config["l10n_repo_url"]))))
+                value["pin"] = False
+            revision_dict[locale] = value
+        await asyncio.gather(*tasks)
+        for task in tasks:
+            (locale, revision) = task.result()
+            revision_dict[locale]["revision"] = revision
+    elif dashboard_revision_info:
```
Comment (bhearsum) [inline treescript/src/treescript/l10n.py:157]: can we kill this branch of the code?

**Review by bhearsum — APPROVED** [review]
This looks fine. Do we need any patches to central, beta, or release to pin some locales before this lands?

File: treescript/src/treescript/l10n.py (hunk 4 above)
Comment (escapewindow) [reply treescript/src/treescript/l10n.py:157]: I think beta and esr68 will use this branch of code til a) central-> beta merge day, and b) we turn off the bumper in esr68. Once those happen, then yes, I'd love to clean this up.

File: treescript/src/treescript/l10n.py (hunk 3 above)
Comment (escapewindow) [reply treescript/src/treescript/l10n.py:102]: As long as we have `tipsonly=1` in the query args, that appears to be true.

File: treescript/src/treescript/l10n.py (hunk 2 above)
Comment (escapewindow) [reply treescript/src/treescript/l10n.py:97]: > This looks fine. Do we need any patches to central, beta, or release to pin some locales before this lands?

Thanks! No, pinning will likely be manual once we determine the default revision for a locale is bad, and we're unable to fix the locale sooner.

**[Edit]** Er, this comment shouldn't have gone into this thread.

Comment (escapewindow) [reply treescript/src/treescript/l10n.py:97]: No, no reason.

Comment (escapewindow) [reply treescript/src/treescript/l10n.py:97]: Hm, I was going to port over `scriptworker.util.load_json_or_yaml_from_url` but that downloads the file and loads it from a tempfile as well. I'll probably leave this for another patch if we want to optimize.

File: treescript/src/treescript/l10n.py (hunk 1 above)
Comment (escapewindow) [reply treescript/src/treescript/l10n.py:96]: Hm. I tried this, and it went back to using `default` for every revision. I haven't dug into why. I also have some vague memory there may be a `format` elsewhere that the `%()s` strings avoid.

Tl;dr, I'm probably going to go back to %s because that was working =\
